File: node/deliberation/src/specialists/equities.mjs
/**
 * @accretive-substrate/deliberation/specialists/equities
 *
 * Equities specialist. Reads sector heatmap + price/volume state for
 * Alpaca-venue symbols.
 *
 * Verdict rules:
 *   sector_strength > 0.7 + symbol_rs > 1.0    → buy
 *   sector_strength < 0.3 OR drawdown > 5%      → sell
 *   market_closed                              → abstain (no edge during halt)
 *   else                                       → hold
 *
 * Domain: 'equities' (venue=alpaca).
 */

import { makeSpecialist, voteFor, abstainBecause } from './base.mjs';

export const EQUITIES_SPECIALIST_NAME = 'equities';

async function argue(pkg) {
  const venue = String(pkg?.trigger?.venue || '').toLowerCase();
  if (venue !== 'alpaca') {
    return abstainBecause(EQUITIES_SPECIALIST_NAME, `venue=${venue} out of scope`);
  }

  const market = pkg?.market_state || {};
  if (market.session === 'closed') {
    return abstainBecause(EQUITIES_SPECIALIST_NAME, 'equity market closed — no decision');
  }

  const sectorStrength = Number(market.sector_strength);
  const rs = Number(market.symbol_rs);
  const drawdown = Number(market.drawdown_pct);

  const citations = [{ type: 'signal', source: 'equities', market }];

  if (!Number.isFinite(sectorStrength) && !Number.isFinite(rs)) {
    return abstainBecause(EQUITIES_SPECIALIST_NAME, 'no sector/RS data in evidence');
  }

  if (sectorStrength > 0.7 && rs > 1.0) {
    return voteFor(
      EQUITIES_SPECIALIST_NAME, 'buy', Math.min(0.85, 0.6 + sectorStrength * 0.2),
      `sector strong ${sectorStrength.toFixed(2)} + RS ${rs.toFixed(2)}`,
      citations,
    );
  }

  if (sectorStrength < 0.3 || drawdown > 5) {
    return voteFor(
      EQUITIES_SPECIALIST_NAME, 'sell', Math.min(0.85, 0.6 + (drawdown > 5 ? drawdown / 20 : 0.1)),
      `sector weak ${sectorStrength.toFixed(2)} or drawdown ${drawdown.toFixed(1)}%`,
      citations,
    );
  }

  return voteFor(
    EQUITIES_SPECIALIST_NAME, 'hold', 0.5,
    `sector ${sectorStrength.toFixed(2)} RS ${Number.isFinite(rs) ? rs.toFixed(2) : 'n/a'} — neutral`,
    citations,
  );
}

export const equitiesSpecialist = makeSpecialist(
  EQUITIES_SPECIALIST_NAME,
  ['equities'],
  argue,
);

File: go/deliberation/pkg/deliberation/specialists/equities/equities.go
// Package equities — Go port of @accretive-substrate/deliberation/specialists/equities.
//
// Filters venue=alpaca. Reads sector_strength + symbol_rs + drawdown_pct
// + session from MarketState.
package equities

import (
	"fmt"
	"math"
	"strings"

	"github.com/ramene/accretive-substrate/go/deliberation/pkg/deliberation"
)

const Name = "equities"

type equityMarket struct {
	Session         string  `json:"session"`
	SectorStrength  float64 `json:"sector_strength"`
	SymbolRS        float64 `json:"symbol_rs"`
	DrawdownPct     float64 `json:"drawdown_pct"`
}

func Argue(pkg deliberation.EvidencePackage) deliberation.Voice {
	venue := strings.ToLower(pkg.Trigger.Venue)
	if venue != "alpaca" {
		return deliberation.AbstainBecause(Name, fmt.Sprintf("venue=%s out of scope", venue))
	}

	var market equityMarket
	_ = deliberation.DecodeMarketStateFor(pkg.MarketState, pkg.Trigger.Symbol, &market)

	if market.Session == "closed" {
		return deliberation.AbstainBecause(Name, "equity market closed — no decision")
	}

	citations := []deliberation.Citation{{
		Type:   "signal",
		Source: "equities",
	}}

	if market.SectorStrength == 0 && market.SymbolRS == 0 {
		return deliberation.AbstainBecause(Name, "no sector/RS data in evidence")
	}

	if market.SectorStrength > 0.7 && market.SymbolRS > 1.0 {
		conf := math.Min(0.85, 0.6+market.SectorStrength*0.2)
		return deliberation.VoteFor(Name, deliberation.VerdictBuy, conf,
			fmt.Sprintf("sector strong %.2f + RS %.2f", market.SectorStrength, market.SymbolRS),
			citations)
	}

	if market.SectorStrength < 0.3 || market.DrawdownPct > 5 {
		bump := 0.1
		if market.DrawdownPct > 5 {
			bump = market.DrawdownPct / 20
		}
		conf := math.Min(0.85, 0.6+bump)
		return deliberation.VoteFor(Name, deliberation.VerdictSell, conf,
			fmt.Sprintf("sector weak %.2f or drawdown %.1f%%", market.SectorStrength, market.DrawdownPct),
			citations)
	}

	rsLabel := "n/a"
	if market.SymbolRS != 0 {
		rsLabel = fmt.Sprintf("%.2f", market.SymbolRS)
	}
	return deliberation.VoteFor(Name, deliberation.VerdictHold, 0.5,
		fmt.Sprintf("sector %.2f RS %s — neutral", market.SectorStrength, rsLabel),
		citations)
}

// Specialist — exported safe-wrapped argue function.
var Specialist = deliberation.MakeSafe(Name, Argue)
